test/module/irohad/subscription/subscription_test.cpp

```diff
@@ -0,0 +1,655 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "main/subscription.hpp"
+
+#include <gtest/gtest.h>
+#include <vector>
+
+#include "module/irohad/subscription/subscription_mocks.hpp"
+
+using namespace iroha;
+
+class SubscriptionTest : public ::testing::Test {
+ public:
+  template <uint32_t ThreadsCount>
+  auto createSubscriptionManager() {
+    using Manager = subscription::SubscriptionManager<ThreadsCount, 1u>;
+    return std::make_shared<Manager>();
+  }
+
+  template <uint32_t Tid,
+            uint64_t Event,
+            typename EventData,
+            typename ObjectType,
+            typename Manager,
+            typename F>
+  auto createSubscriber(std::shared_ptr<Manager> const &manager,
+                        ObjectType &&initial,
+                        F &&f) {
+    using Dispatcher = typename Manager::Dispatcher;
+    using Subscriber = subscription::
+        SubscriberImpl<uint64_t, Dispatcher, ObjectType, EventData>;
+
+    auto subscriber = std::make_shared<Subscriber>(
+        manager->template getEngine<uint64_t, EventData>(), std::move(initial));
+
+    subscriber->setCallback(
+        [f(std::forward<F>(f))](
+            auto, ObjectType &obj, uint64_t key, EventData data) mutable {
+          ASSERT_EQ(key, Event);
+          std::forward<F>(f)(obj, std::move(data));
+        });
+    subscriber->template subscribe<Tid>(0, Event);
+    return subscriber;
+  }
+
+  using MockDispatcher = subscription::MockDispatcher;
+  using MockSubscriber =
+      subscription::MockSubscriber<uint32_t, MockDispatcher, std::string>;
+  using TestEngine = subscription::SubscriptionEngine<
+      uint32_t,
+      MockDispatcher,
+      subscription::Subscriber<uint32_t, MockDispatcher, std::string>>;
+
+  auto createMockSubscriber(std::shared_ptr<TestEngine> const &engine) {
+    return std::make_shared<MockSubscriber>(engine);
+  }
+
+  auto createTestEngine(std::shared_ptr<MockDispatcher> const &dispatcher) {
+    return std::make_shared<TestEngine>(dispatcher);
+  }
+
+  auto createDispatcher() {
+    return std::make_shared<MockDispatcher>();
+  }
+};
+
+/**
+ * @given subscription engine
+ * @when subscriber is present
+ * @and notification is called
+ * @then subscriber must receive correct data from notification
+ */
+TEST_F(SubscriptionTest, SimpleExecutionTest) {
+  auto manager = createSubscriptionManager<1>();
+  utils::WaitForSingleObject complete;
+
+  std::string test_value = "the fast and the furious";
+  auto subscriber = createSubscriber<0ul, 1ull, std::string>(
+      manager, false, [&complete, test_value](auto, std::string value) {
+        ASSERT_EQ(test_value, value);
+        complete.set();
+      });
+
+  manager->notify(1ull, test_value);
+  ASSERT_TRUE(complete.wait(std::chrono::minutes(1ull)));
+
+  manager->dispose();
```

Comment:
Can it be disposed in destructor? Or if it's not required, maybe omit this call from tests?

Reply:
I can't dispose in destructor because tasks can make also some subscriptions ```lock```ing manager in strong pointer. It means that this pointer can be the last reference to manager. And leaving the scope there will be attempt to stop and join threads from one of the handlers.